snappy-core/src/main/scala/org/apache/spark/sql/execution/joins/LocalJoin.scala

```diff
@@ -66,48 +73,108 @@ case class LocalJoin(leftKeys: Seq[Expression],
     }
     val numOutputRows = longMetric("numOutputRows")
 
-    narrowPartitions(buildPlan.execute(), streamedPlan.execute(), true) {
-      (buildIter, streamIter) => {
-        val hashed = HashedRelation(buildIter, numBuildRows, buildSideKeyGenerator)
+    val h1 = (buildIter: Iterator[InternalRow]) => Iterator[HashedRelation] {
+      HashedRelation(buildIter, numBuildRows, buildSideKeyGenerator)
+    }
+
+    val buildRDD = buildPlan.execute()
+    val streamRDD = streamedPlan.execute()
+
+    val sc = buildRDD.sparkContext
+    val hashedRDD = new HashRelationRDD(sc, buildRDD,
+      streamRDD.partitions.length, sc.clean(h1))
+
+    narrowPartitions(hashedRDD, streamRDD, true) {
+      (hashedIter, streamIter) => {
+        val hashed = hashedIter.next()
         hashJoin(streamIter, numStreamedRows, hashed, numOutputRows)
       }
     }
   }
 
-  def narrowPartitions(buildRDD: RDD[InternalRow], streamRDD: RDD[InternalRow], preservesPartitioning: Boolean)
-      (f: (Iterator[InternalRow], Iterator[InternalRow]) => Iterator[InternalRow]): NarrowPartitionsRDD = {
-    val sc = buildRDD.sparkContext
+  def narrowPartitions(hashedRDD: RDD[HashedRelation], streamRDD: RDD[InternalRow], preservesPartitioning: Boolean)
+      (f: (Iterator[HashedRelation], Iterator[InternalRow]) => Iterator[InternalRow]): NarrowPartitionsRDD = {
+    val sc = hashedRDD.sparkContext
     new NarrowPartitionsRDD(sc, sc.clean(f),
-      buildRDD, streamRDD, preservesPartitioning)
+      hashedRDD, streamRDD, preservesPartitioning)
   }
 }
 
+// Helper object jut to take a global sync across tasks.
+// Furture helper methods can come here
+private[spark] object HashRelationRDD
+
+private[spark] class HashRelationRDD(
+    sc: SparkContext,
+    var buildRDD: RDD[InternalRow],
+    val maxPartitions: Int,
+    var f: (Iterator[InternalRow]) => Iterator[HashedRelation]
+    ) extends RDD[HashedRelation](sc, Seq(new OneToOneDependency(buildRDD))) {
+
+  override def compute(s: Partition, context: TaskContext): Iterator[HashedRelation] = {
+
+    val blockId = RDDBlockId(this.id, s.index)
+
+    val rel1 = SparkEnv.get.blockManager.getSingle(blockId) match {
+      case Some(x) => x.asInstanceOf[HashedRelation]
+      case None => {
+        HashRelationRDD.synchronized {
```

Comment:
Also I think other cores should be busy in processing other parts of the query (or not)? [Rishi] I need to change the order of iterator , so that scanning of partition table will be first. Then the hashed relation. This will allow threads to go parallel without waiting. First thread finishing partitioned scan will also scan the replicated table and cache it. Others when finished with their scanning need to wait till HashedRelation is available.